internal/subscription/validation.go

```diff
@@ -114,78 +155,206 @@ func GetValidatedModelForUpdate(ctx context.Context, req *metrov1.Subscription)
+			Credentials:  creds,
+		}
+	}
+
+	return nil
+}
+
+func getValidatedAckDeadline(_ context.Context, m *Model, req *metrov1.Subscription) error {
+	ackDeadlineSeconds := req.AckDeadlineSeconds
+
+	if ackDeadlineSeconds == 0 {
+		m.AckDeadlineSeconds = DefaultAckDeadlineSeconds
+		return nil
 	}
-	return "", nil
+
+	if ackDeadlineSeconds < MinAckDeadlineSeconds {
+		return ErrInvalidAckDeadline
+	}
+
+	if ackDeadlineSeconds > MaxAckDeadlineSeconds {
+		return ErrInvalidAckDeadline
+	}
+
+	m.AckDeadlineSeconds = ackDeadlineSeconds
+	return nil
+}
+
+func validateRetryPolicy(_ context.Context, m *Model, req *metrov1.Subscription) error {
+	retryPolicy := req.GetRetryPolicy()
+
+	if retryPolicy == nil {
+		m.RetryPolicy = &RetryPolicy{
+			MinimumBackoff: DefaultMinBackOffSeconds,
+			MaximumBackoff: DefaultMaxBackoffSeconds,
+		}
+		return nil
+	}
+
+	// Note that input backoff is of type durationpb.Duration. below we are using only seconds part
+	// of the struct, for now we are ignoring the nanoseconds part since metro supports only integer
+	// values for backoff
+	if retryPolicy.MinimumBackoff.Seconds > retryPolicy.MaximumBackoff.Seconds {
+		return ErrInvalidMinAndMaxBackoff
+	}
+
+	if retryPolicy.MinimumBackoff.Seconds < MinBackOffSeconds {
+		return ErrInvalidMinBackoff
+	}
+
+	if retryPolicy.MaximumBackoff.Seconds > MaxBackOffSeconds {
+		return ErrInvalidMaxBackoff
+	}
+
+	m.RetryPolicy = &RetryPolicy{
+		MinimumBackoff: uint(retryPolicy.MinimumBackoff.Seconds),
+		MaximumBackoff: uint(retryPolicy.MaximumBackoff.Seconds),
+	}
+
+	return nil
+}
+
+func validatedDeadLetterPolicy(_ context.Context, m *Model, req *metrov1.Subscription) error {
+	defaultDeadLetterTopic := topic.GetTopicName(m.ExtractedSubscriptionProjectID, m.ExtractedSubscriptionName+topic.DeadLetterTopicSuffix)
+
+	dlpolicy := req.GetDeadLetterPolicy()
+	if dlpolicy == nil {
+		m.DeadLetterPolicy = &DeadLetterPolicy{
+			MaxDeliveryAttempts: DefaultDeliveryAttempts,
+			DeadLetterTopic:     defaultDeadLetterTopic,
+		}
+
+		return nil
+	}
+
+	if dlpolicy.MaxDeliveryAttempts < MinDeliveryAttempts {
+		return ErrInvalidMaxDeliveryAttempt
+	}
+
+	if dlpolicy.MaxDeliveryAttempts > MaxDeliveryAttempts {
+		return ErrInvalidMaxDeliveryAttempt
+	}
+
+	// TODO: check and validate if dltopic is present in input, validate topic is valid topic
+	// and use topic from input instead
+	m.DeadLetterPolicy = &DeadLetterPolicy{
+		MaxDeliveryAttempts: dlpolicy.MaxDeliveryAttempts,
+		DeadLetterTopic:     defaultDeadLetterTopic,
+	}
+
+	return nil
 }
 
 func extractSubscriptionMetaAndValidate(_ context.Context, name string) (projectID string, subscriptionName string, err error) {
 	match := subscriptionNameRegex.FindStringSubmatch(name)
 	if len(match) != 3 {
-		err = fmt.Errorf("invalid subscription name")
-		return "", "", err
+		return "", "", errors.New(fmt.Sprintf("Invalid [subscriptions] name: (name=%s)", name))
 	}
 	projectID = subscriptionNameRegex.FindStringSubmatch(name)[1]
```

Comment:
projectID = match[1]
subscriptionName = match[2]
This would make things more efficient, wouldn't it?

Reply:
Yes, we dont need to parse multiple times